### PR TITLE
Fire - Pauline

### PR DESCRIPTION
**Pull request by ghostfruitleaf**
# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I wrote a custom method to sort works in a certain category by votes, then alphabetically. 
Describe how you approached testing that model method. What edge cases did you come up with? | I started with verifying that all objects found in the nominal case of picking the work with the most votes have the same category and are the only two objects in the small set of test works with said category. Then I wrote a test for the following situations: 1) Most votes (nominal), 2) Top works have same number of votes, pick work that shows up earlier in the alphabet (edge) -- This was similar for the Work.spotlight method, except since works of different categories can share the same title, 
What are `session` and `flash`? What is the difference between them? | Session and flash describe a state you can put your app in that is temporary. Session, however, lasts until a browser's cookies are cleared or the user logs out (after having logged into a session prior), and allows during that time certain views and action an "authenticated" user can access. However, flash only lasts for one request cycle, saving certain messages and information to display between request cycles and disappearing after the next in a "flash hash." This means if your previous request lands you on a new page and has info added to its flash hash, on that new page it can display information in the flash hash that disappears if the user refreshes, closes the browser, or visits another page (whereas being logged in or out of a session doesn't change whether you can get flash messages, rather which ones will be displayed to you depending on authorizations.)
What was one thing that you gained more clarity on through this assignment? | I learned a lot about how to both create and display flash messages and got practice being my own scrum lead with kanban. I also became more comforting writing rails tests through the custom methods.
What is the Heroku URL of your deployed application? | https://media-ranker-replica.herokuapp.com/

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class ApplicationController < ActionController::Base
+  def find_current_user
+    @user = session[:user_id].nil? ? nil : User.find_by_id(session[:user_id])
```
Comment (tildeee) [inline app/controllers/application_controller.rb:3]: Slick ternary!

File: test/models/user_test.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    it "can get information of works the user voted for" do
+      # Assert
+      expect(users(:user).works.count).must_equal 2
+      # check info
+      expect(users(:user).works.find_by(title: "The Book")).must_equal works(:book)
+      expect(users(:user).works.find_by(title: "A")).must_equal works(:workAA)
+      users(:user).works.each do |work|
+        expect(work).must_be_instance_of Work
+      end
+    end
```
Comment (tildeee) [inline test/models/user_test.rb:38]: This is a good test and i like where you're going with it, but this is veering into the territory of "testing methods that should be tesetd somewhere else". For example, as long as we can assume that we have a valid array of `Work`s, this test can feel like it's testing `Work` functionality.

Nothing to start or stop doing imo, just worth thinking about

File: test/models/vote_test.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      # confirms that object exists and fields match
+      expect(votes(:vote).work).must_equal works(:book)
+      expect(votes(:vote).work.title).must_equal works(:book).title
+      expect(votes(:vote).work.category).must_equal works(:book).category
+      expect(votes(:vote).work_id).must_equal works(:book).id
+    end
+  end
+
+  describe 'validations' do
+    it 'prevents multiple votes for a work by the same user' do
```
Comment (tildeee) [inline test/models/vote_test.rb:33]: Well written test and well-titled test!! I love how the code in this test reads

File: test/models/work_test.rb [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+  describe "custom methods" do
+    before do
+      @user = User.create(username: 'voter')
+      @newer_work = Work.new(category: 'movie', title: 'A')
+    end
+
+    describe "spotlight" do
+      before do
+        Vote.delete_all # for this particular section we need to manipulate votes a certain way
```
Comment (tildeee) [inline test/models/work_test.rb:130]: Nice! This is great, and the comment helps me understand btw!

File: test/models/work_test.rb [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+      end
+      it "returns work with most votes" do
+        # @work1.save
+        # @work2.save
+
+        Vote.create(work_id: works(:workBB).id, user_id: @user.id)
+
+        expect(Work.spotlight).must_equal works(:workBB)
+        expect(Work.spotlight.votes.count).must_equal 1
+        expect(Work.spotlight.votes.count).must_equal works(:workBB).votes.count
```
Comment (tildeee) [inline test/models/work_test.rb:144]: Going off of the longer comment from before, since we've tested that `Work`s indeed have votes, this specific line might feel like a redundant test-- it's testing that `votes.count` of two arrays is the same.

If we really want to test that a vote count increases, this might feel more like a controller test-- testing that the action of voting increases the vote count.

File: test/models/work_test.rb [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+      end
+      it "returns work earlier in alphabet if top have same number of votes" do
+
+        expect(Work.spotlight).must_equal works(:workBA)
+        expect(Work.spotlight.votes.count).must_equal 0
+        expect(Work.spotlight.votes.count).must_equal works(:workBB).votes.count
+      end
+
+      it "returns work updated more recently if same title (possible with diff. cat) and same number of votes" do
```
Comment (tildeee) [inline test/models/work_test.rb:153]: Great test!

File: test/models/work_test.rb [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        work_hash = Work.work_hash
+        expect(work_hash.length).must_equal 3
+
+        total_works = 0
+        work_hash.each do |cat, array|
+          expect(array.length).must_equal Work.where(category: cat.to_s.chop).count
+          total_works += array.length
+          array.each {|work| expect(work.category).must_equal cat.to_s.chop}
+        end
+        expect(total_works).must_equal Work.all.count
```
Comment (tildeee) [inline test/models/work_test.rb:220]: Cooool test!!!!! I would probably say that the `total_works` bit is a little redundant with the looped `expect(array.length)...` assert. I'd feel comfortable if you kept the loop and dropped the `total_works` bit!

**Review by tildeee — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️ -- Clicking on the "Logged in as ..." seems to have a bug on the deployed version?
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️?
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️?
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 

### Additional Feedback

Great work on this project, Pauline!!
Since we talked about tests and level of thoroughness in tests this morning, I've made some comments on a few tests that I think were a little redundant.

In general, the project code looks great. The deployed version looks so nice! alsO [JUST ONE BEAT. COMMUNI-CA-TIOOON](https://www.youtube.com/watch?v=Tf40QBWLRVE)

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | 
| Logical/Organized | ✅
